### PR TITLE
[DF] add Davix vfs module to sqlite ds

### PR DESCRIPTION
**Pull request by jblomer**
Allows for direct processing of sqlite files under HTTP(S) URLs provided that Davix is available.

Replaces #2578

I think it would be nice to have a unit test for reading of remote files.  Unfortunately, Davix does not understand the `http:///path/to/file` trick to use a URL on local paths.  Can we upload tree/dataframe/test/RSqliteDS_test.sqlite to https://root.cern.ch/... for testing?

### CODE REVIEW

**Comment by bluehood** [comment]
@phsft-bot build

**Comment by bluehood** [comment]
@jblomer seems that there's a problem with one of the changes in dataframe's CMakeLists:
```
09:31:20 CMake Error at tree/dataframe/CMakeLists.txt:34 (target_include_directories):
09:31:20   target_include_directories called with invalid arguments
```

**Comment by bluehood** [comment]
this is the current issue:
```
[RUN] RSqliteDS.Davix
Error in <VfsRdOnlyOpen>: Failure (Neon): Server certificate verification failed: issuer is not trusted after 3 attempts
```

File: tree/dataframe/src/RSqliteDS.cxx [hunk 1]
```diff
@@ -60,11 +345,24 @@ constexpr char const *RSqliteDS::fgTypeNames[];
 RSqliteDS::RSqliteDS(std::string_view fileName, std::string_view query)
    : fDb(nullptr), fQuery(nullptr), fNSlots(0), fNRow(0)
 {
+   static bool isRegistered = Register();
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: the `static` here means that only the first `RSqliteDS` constructor call will try to register...if that fails, all subsequent ctor calls will not even try to register and just throw. does it not make sense to have a retry mechanism?

**Comment by Axel-Naumann** [comment]
The only urls I see are at `https://root.cern.ch/`. I don't see how that could possibly fail - the cert is signed by Comodo which is known / accepted by "the world". If we don't solve this we can still revert to  `http://`: `/files/` is one of the few directories also served as `http`.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 2]
```diff
@@ -13,19 +13,304 @@
...
+};
+
+////////////////////////////////////////////////////////////////////////////
+/// Releases the resources associated to a file opened with davix
+static int VfsRdOnlyClose(sqlite3_file *pFile)
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   if (p->pos.close(p->fd, &err) == -1) {
+      p->~VfsRootFile();
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: would `delete p` work too? manually calling destructors looks weird :sweat_smile:

File: tree/dataframe/src/RSqliteDS.cxx [hunk 3]
```diff
@@ -13,19 +13,304 @@
...
+   if (p->pos.pread(p->fd, zBuf, count, offset, &err) == -1) {
+      return SQLITE_IOERR;
+   }
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyWrite(sqlite3_file * /*pFile*/, const void * /*zBuf*/, int /*iAmt*/, sqlite_int64 /*iOfst*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: as a non-sqlite expert, I really need a comment in this function and the following to understand what's going on :)

File: tree/dataframe/src/RSqliteDS.cxx [hunk 4]
```diff
@@ -13,19 +13,304 @@
...
+   if (p->pos.stat(NULL, zName, &buf, NULL) == -1) {
+      return SQLITE_IOERR;
+   }
+   p->size = buf.st_size;
+
+   p->pFile.pMethods = &io_methods;
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyDelete(sqlite3_vfs *__attribute__((unused)), const char * /*zName*/, int /*syncDir*/)
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: why does _this particular_ argument require a `__attribute__((unused))` (but not all others)?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 5]
```diff
@@ -13,19 +13,304 @@
...
+namespace {
+
+constexpr char const *gSQliteVfsName = "ROOT-Davix-readonly";
+
+#ifdef R__HAS_DAVIX
+
+/**
+ * The "derived class" from sqlite3_file that gets passed to the file operation callbacks.
+ */
+struct VfsRootFile {
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: we need a long-ish explanation of how this inheritance with no inheritance work :) c++ developers won't know

File: tree/dataframe/src/RSqliteDS.cxx [hunk 6]
```diff
@@ -13,19 +13,304 @@
...
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Fills a new VfsRootFile struct enclosing a Davix file
+static int VfsRdOnlyOpen(sqlite3_vfs * /*vfs*/, const char *zName, sqlite3_file *pFile, int flags, int * /*pOutFlags*/)
+{
+   VfsRootFile *p = new (pFile) VfsRootFile();
+   p->pFile.pMethods = nullptr;
+
+   static const sqlite3_io_methods io_methods = {
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: please comment :) I can guess what's going on because I once glimpsed into the abyss of the cpython C api :P

File: tree/dataframe/src/RSqliteDS.cxx [hunk 7]
```diff
@@ -132,10 +430,10 @@ RSqliteDS::RSqliteDS(std::string_view fileName, std::string_view query)
 RSqliteDS::~RSqliteDS()
 {
    // sqlite3_finalize returns the error code of the most recent operation on fQuery.
-   (void) sqlite3_finalize(fQuery);
+   (void)sqlite3_finalize(fQuery);
```
Comment (bluehood) [inline tree/dataframe/src/RSqliteDS.cxx]: is the `(void)` cast needed? i.e. do these kind of ignored return values trigger compiler warnings? (can't see it e.g. [here](https://godbolt.org/z/La9zRq))

File: tree/dataframe/CMakeLists.txt [hunk 8]
```diff
@@ -37,6 +37,7 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTDataFrame
                               SOURCES ${sources}
                               DICTIONARY_OPTIONS "-writeEmptyRootPCM"
                               LIBRARIES ${TBB_LIBRARIES} ${ARROW_SHARED_LIB} ${DATAFRAME_SQLITE_LIBRARIES}
+			      ${SQLITE_LIBRARIES} Davix::Davix
```
Comment (amadio) [inline tree/dataframe/CMakeLists.txt]: This will break if Davix is disabled. You need to conditionally link to Davix.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 9]
```diff
@@ -14,6 +14,12 @@
     \brief RDataFrame data source class for reading SQlite files.
 */
 // clang-format on
+#include <davix.hpp>
+
+#include "TROOT.h"
+#include "TSystem.h"
+
+#include <string.h>
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: There is already `#include <cstring>` below. Why is this needed?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 10]
```diff
@@ -26,6 +32,328 @@
 #include <cstring>
 #include <stdexcept>
 
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/time.h>
+
+#include <linux/limits.h>
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: This will be a hassle for portability if Davix receives support on Windows. Please try to keep the code portable as much as possible by using include guards for platform-specific headers. Also, please use alphabetical order for includes if you can.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 11]
```diff
@@ -26,6 +32,328 @@
...
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyRead(
+   sqlite3_file *pFile,
+   void *zBuf,
+   int count,
+   sqlite_int64 offset
+) {
+   Davix::DavixError *err = NULL;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: `NULL` is a C thing. Please use `nullptr` instead.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 12]
```diff
@@ -26,6 +32,328 @@
...
+   return SQLITE_OPEN_READONLY;
+}
+
+static int VfsRdOnlyOpen(
+   sqlite3_vfs * /*vfs*/,
+   const char *zName,
+   sqlite3_file *pFile,
+   int flags,
+   int * /*pOutFlags*/)
+{
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: This and many functions above are not following our coding style for the function prototypes.
I also don't see the point in declaring these function `static` if they are already inside an unnamed namespace. Isn't that implied?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 13]
```diff
@@ -26,6 +32,328 @@
...
+      NULL,
+      NULL
+   };
+
+   if (flags & SQLITE_OPEN_READWRITE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_DELETEONCLOSE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_EXCLUSIVE)
+      return SQLITE_IOERR;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: These 3 `if`s could be combined into a single one, no?
```
if (flags & (SQLITE_OPEN_READWRITE | SQLITE_OPEN_DELETEONCLOSE | SQLITE_OPEN_EXCLUSIVE))
   return SQLITE_IOERR;
```

File: tree/dataframe/src/RSqliteDS.cxx [hunk 14]
```diff
@@ -26,6 +32,328 @@
...
+
+static int VfsRdOnlyOpen(
+   sqlite3_vfs * /*vfs*/,
+   const char *zName,
+   sqlite3_file *pFile,
+   int flags,
+   int * /*pOutFlags*/)
+{
+   VfsRootFile *p = new (pFile) VfsRootFile();
+   p->pFile.pMethods = NULL;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Null pointer dereference bug here.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 15]
```diff
@@ -26,6 +32,328 @@
...
+   if (flags & SQLITE_OPEN_DELETEONCLOSE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_EXCLUSIVE)
+      return SQLITE_IOERR;
+
+   Davix::DavixError *err = NULL;
+   p->fd = p->pos.open(NULL, zName, O_RDONLY, &err);
+
+   if (!p->fd) {
+      printf("%s\n", err->getErrMsg().c_str());
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Another possible null pointer dereference bug.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 16]
```diff
@@ -26,6 +32,328 @@
...
+   const char * /*zPath*/,
+   int flags,
+   int *pResOut)
+   {
+      *pResOut = 0;
+      if (flags == SQLITE_ACCESS_READWRITE) {
+         return SQLITE_OPEN_READONLY;
+      }
+   return SQLITE_OK;
+   }
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Coding style isn't followed here too. Indentation is wrong.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 17]
```diff
@@ -26,6 +32,328 @@
...
+
+int VfsRdOnlyFullPathname(
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Why is this not `static`, when all others are `static`? Please add a comment if this is intentional.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 18]
```diff
@@ -26,6 +32,328 @@
...
+     close(fd);
+   }
+  return nBuf;
+}
+
+static int VfsRdOnlySleep(
+   sqlite3_vfs * /*vfs*/,
+   int microseconds)
+{
+   gSystem->Sleep(microseconds / 1000);
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: This will fail to sleep for any number of microseconds smaller than 1000, because the division will give you 0.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 19]
```diff
@@ -26,6 +32,328 @@
...
+   return microseconds;
+}
+
+static int VfsRdOnlyGetLastError(
+   sqlite3_vfs * /*vfs*/,
+   int /*not_used1*/,
+   char * /*not_used2*/)
+{
+   return 0;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Please add a comment why you can always return 0 here. It's not obvious to me that this should never return other value if there is a genuine error somewhere.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 20]
```diff
@@ -26,6 +32,328 @@
...
+ * Taken from unixCurrentTimeInt64()
+ */
+static int VfsRdOnlyCurrentTimeInt64(
+   sqlite3_vfs * /*vfs*/,
+   sqlite3_int64 *piNow)
+{
+   static const sqlite3_int64 unixEpoch = 24405875*(sqlite3_int64)8640000;
+   int rc = SQLITE_OK;
+   struct timeval sNow;
+   if (gettimeofday(&sNow, 0) == 0) {
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: From `man 3 gettimeofday`:
> POSIX.1-2008 marks `gettimeofday()` as obsolete, recommending the use of `clock_gettime(2)` instead.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 21]
```diff
@@ -26,6 +32,328 @@
...
+}
+
+static bool IsURL(std::string_view fileName) {
+   auto haystack = std::string(fileName);
+   if (haystack.find ("http://")==0)
+      return true;
+   if (haystack.find ("https://")==0)
+      return true;
+   return false;
+}
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Please consider using `compare()` here instead of `find()`.
https://en.cppreference.com/w/cpp/string/basic_string/compare

File: tree/dataframe/src/RSqliteDS.cxx [hunk 22]
```diff
@@ -60,11 +388,22 @@ constexpr char const *RSqliteDS::fgTypeNames[];
...
    int retval;
 
-   retval = sqlite3_open_v2(std::string(fileName).c_str(), &fDb, SQLITE_OPEN_READONLY | SQLITE_OPEN_NOMUTEX, nullptr);
-   if (retval != SQLITE_OK)
-      SqliteError(retval);
+   // Opening the layer
+   if (IsURL(fileName)) {
+      if (!isRegistered)
+         throw std::runtime_error("Error");
+      retval = sqlite3_open_v2(std::string(fileName).c_str(), &fDb, SQLITE_OPEN_READONLY | SQLITE_OPEN_NOMUTEX, gSQliteVfsName);
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Please don't use `string_view` if the contents will be copied anyway.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 23]
```diff
@@ -26,6 +32,328 @@
...
+   NULL, /* app data */
+   VfsRdOnlyOpen,
+   VfsRdOnlyDelete,
+   VfsRdOnlyAccess,
+   VfsRdOnlyFullPathname, // to do
+   NULL,
+   NULL,
+   NULL,
+   NULL,
+   VfsRdOnlyRandomness, // eventually --> TODO TRandom calls
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Should the TODO's be addressed as part of the PR?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 24]
```diff
@@ -26,6 +32,328 @@
...
+static int VfsRdOnlySync(
+   sqlite3_file * /*pFile*/,
+   int /*flags*/)
+{
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyFileSize(sqlite3_file *pFile, sqlite_int64 *pSize) {
+  VfsRootFile *p = reinterpret_cast<VfsRootFile*>(pFile);
+  *pSize = p->size;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Null pointer dereference.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 25]
```diff
@@ -26,6 +32,328 @@
...
+
+static int VfsRdOnlyRead(
+   sqlite3_file *pFile,
+   void *zBuf,
+   int count,
+   sqlite_int64 offset
+) {
+   Davix::DavixError *err = NULL;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile*>(pFile);
+   if (p->pos.pread(p->fd, zBuf, count, offset, &err) == -1) {
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Null pointer dereference.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 26]
```diff
@@ -26,6 +32,328 @@
...
+   DAVIX_FD* fd;
+   uint64_t size;
+   Davix::Context c;
+   Davix::DavPosix pos;
+};
+
+static int VfsRdOnlyClose(sqlite3_file *pFile) {
+   Davix::DavixError *err = NULL;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile*>(pFile);
+   if (p->pos.close(p->fd, &err) == -1){
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Null pointer dereference.

**Comment by jblomer** [comment]
@bluehood @Axel-Naumann The HTTPS error is specific to macOS. As far as I can tell, Davix resp. libNeon does not find the default CAs on Mac.  The same problem is for ROOT files, e.g.

```
TFile *f = TFile::Open("https://root.cern.ch/files/h1big.root");
```

works on Linux but not on Mac. For this test, I could ifdef the URLs to HTTP on Mac. But perhaps we should rather solve the underlying CA store issue on Mac.

File: tree/dataframe/src/RSqliteDS.cxx (hunk 1 above)
Comment (jblomer) [reply tree/dataframe/src/RSqliteDS.cxx]: I think we wouldn't gain from a retry. The registration is only setting the function pointers. It always works iff Davix is available. For the actual HTTP operations, Davix implements a retry logic.

Comment (bluehood) [reply tree/dataframe/src/RSqliteDS.cxx]: I see, but then this is not really `isRegistered` but rather `isDavixAvailable`..?

File: tree/dataframe/src/RSqliteDS.cxx (hunk 12 above)
Comment (jblomer) [reply tree/dataframe/src/RSqliteDS.cxx]: I removed the `static` keyword from the function definitions. Regarding the prototype, as the functions are callback functions, the prototype is imposed by the sqlite library.

File: tree/dataframe/src/RSqliteDS.cxx (hunk 14 above)
Comment (jblomer) [reply tree/dataframe/src/RSqliteDS.cxx]: Sqlite ensures that the input pointer `p` is not `NULL`.

File: tree/dataframe/src/RSqliteDS.cxx (hunk 22 above)
Comment (jblomer) [reply tree/dataframe/src/RSqliteDS.cxx]: I changed it but I wonder if you are sure about this. It makes the interface slightly inconsistent (one parameter is a `string`, the other one a `string_view`) and the interface becomes dependent on the implementation details.

Comment (bluehood) [reply tree/dataframe/src/RSqliteDS.cxx]: there are diverging opinions on this in the team. Philippe (and maybe Axel) is very much in favor of always exposing string_view in user-facing APIs, I'm somewhat undecided, Guilherme is very much against it especially when it results in strings being constructed in the body of the function.

in this case I don't think it really matter as it's an internal function (datasources are always constructed by factory functions, which incidentally imo should take `string_view`s)

Comment (pcanal) [reply tree/dataframe/src/RSqliteDS.cxx]: > when it results in strings being constructed in the body of the function.

Indeed if/when the input is copied *as is* then it make senses, otherwise having the argument being an std::string makes a useless/unnecessary copy.

Comment (pcanal) [reply tree/dataframe/src/RSqliteDS.cxx]: In this context, the 'efficient' solution is to have 3 overload (const char*, const std::string& and std::string_view) and to forward the later two to the first one [The blocker here is that you end up calling a C interface and thus you need a C-style string but you do *not* need a copy of the string.]

Comment (pcanal) [reply tree/dataframe/src/RSqliteDS.cxx]: If/since it is an internal routine it might make sense to 'just' use "const char*" and do the necessary conversion upstream.

File: tree/dataframe/src/RSqliteDS.cxx (hunk 14 above)
Comment (amadio) [reply tree/dataframe/src/RSqliteDS.cxx]: I see, `p` comes from a call to placement new on `pFile`. I think you meant `pFile` cannot be null?

Comment (jblomer) [reply tree/dataframe/src/RSqliteDS.cxx]: Yes

File: tree/dataframe/src/RSqliteDS.cxx (hunk 12 above)
Comment (jblomer) [reply tree/dataframe/src/RSqliteDS.cxx]: As for the code style, this must have been picked up from older commits and is meanwhile resolved.

File: tree/dataframe/src/RSqliteDS.cxx (hunk 22 above)
Comment (amadio) [reply tree/dataframe/src/RSqliteDS.cxx]: Let me just explain my view. I am against `string_view` because it creates problems when it's used in user interfaces and can be either our backported implementation or the standard one, depending on configuration flags. This is a point where we can break ABI (or even API if our backport deviates from the standard, which it did for a while). It's a similar problem to using SIMD types in the interfaces, which I also strongly oppose. Using `string_view` would not be a problem if our baseline was C++17, but since it's C++11, it is. In most places where `string_view` has been used, it brings basically no benefits, because its contents are copied anyway. It's in those places I'm advocating for not using it. I think that we've had many more cons than pros since the introduction of this backport into ROOT: https://root-forum.cern.ch/search?q=string_view. That said, ultimately the choice is of the implementer, so take my comment above (the first one) only as a suggestion.

Comment (Axel-Naumann) [reply tree/dataframe/src/RSqliteDS.cxx]: You are discussing a design pattern that is beyond this PR. To discuss it, please send an email to rootdev@cern.ch instead.

ROOT must be built and used (as includes and libraries) with a consistent configuration; the C++ standard is part of this and people get it from `root-config --cflags`. I understand that you don't like that, but let's discuss that at rootdev. Thanks!

**Comment by jblomer** [comment]
I've been looking into the failure on macOS with @amadio and it appears to be a problem with the build node.

File: tree/dataframe/CMakeLists.txt [hunk 27]
```diff
@@ -36,8 +27,20 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTDataFrame
...
                               DEPENDENCIES Tree TreePlayer Hist RIO ROOTVecOps Imt
                               ${TREEPLAYER_DEPENDENCIES})
+if (ARROW_FOUND)
+  target_include_directories(ROOTDataFrame PRIVATE ${ARROW_INCLUDE_DIR})
+  target_link_libraries(ROOTDataFrame PRIVATE ${ARROW_SHARED_LIB})
+endif()
+
+if (sqlite)
+  target_include_directories(ROOTDataFrame PRIVATE ${SQLITE_INCLUDE_DIR})
+  target_link_libraries(ROOTDataFrame PRIVATE ${SQLITE_LIBRARIES})
```
Comment (amadio) [inline tree/dataframe/CMakeLists.txt]: Shouldn't these be `PUBLIC` since `sqlite3.h` is included in `ROOT/RSqliteDS.hxx`, which is included by users? Same is probably true for the arrow equivalents above and maybe for `Davix::Davix` below if `davix.hpp` is included in a header users need.

File: tree/dataframe/src/RSqliteDS.cxx [hunk 28]
```diff
@@ -13,19 +13,352 @@
...
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   auto retval = p->pos.close(p->fd, &err);
+   // We can't use delete because the storage for p is managed by sqlite
+   p->~VfsRootFile();
+   if (retval == -1) {
+      return SQLITE_IOERR_CLOSE;
+   }
+   return SQLITE_OK;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: `return retval == -1 ? SQLITE_IOERR_CLOSE : SQLITE_OK;`?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 29]
```diff
@@ -13,19 +13,352 @@
...
+
+////////////////////////////////////////////////////////////////////////////
+/// Issues an HTTP range request to read a chunk from a remote file
+int VfsRdOnlyRead(sqlite3_file *pFile, void *zBuf, int count, sqlite_int64 offset)
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   if (p->pos.pread(p->fd, zBuf, count, offset, &err) == -1) {
+      return SQLITE_IOERR;
+   }
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: Why no `retval` (i.e. different style than above)?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 30]
```diff
@@ -13,19 +13,352 @@
...
+int VfsRdOnlyUnlock(sqlite3_file * /*pFile*/, int /*level*/)
+{
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// As the database is read-only, locks for concurrent access are no-ops and always succeeds
+int VfsRdOnlyCheckReservedLock(sqlite3_file * /*pFile*/, int *pResOut)
+{
+   *pResOut = 0;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: `nullptr`?

File: tree/dataframe/src/RSqliteDS.cxx [hunk 31]
```diff
@@ -13,19 +13,352 @@
...
+int VfsRdOnlyDelete(sqlite3_vfs * /*vfs*/, const char * /*zName*/, int /*syncDir*/)
+{
+   return SQLITE_IOERR_DELETE;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Access control always allows read-only access to databases
+int VfsRdOnlyAccess(sqlite3_vfs * /*vfs*/, const char * /*zPath*/, int flags, int *pResOut)
+{
+   *pResOut = 0;
```
Comment (amadio) [inline tree/dataframe/src/RSqliteDS.cxx]: `nullptr`?

File: tree/dataframe/src/RSqliteDS.cxx (hunk 30 above)
Comment (Axel-Naumann) [reply tree/dataframe/src/RSqliteDS.cxx]: No, he's setting the integer (pointed to by `pResOut`) to the value `0`. Same for the others.

**Comment by Axel-Naumann** [comment]
Could you check with our MacOS node people (I think Radu)?

File: tree/dataframe/CMakeLists.txt (hunk 27 above)
Comment (jblomer) [reply tree/dataframe/CMakeLists.txt]: The header of the sqlite data source is now changed such that it doesn't include `sqlite3.h` anymore. The arrow data source header didn't include anything special in the first place.

**Comment by jblomer** [comment]
Coming back to the HTTPS issue: together with @radupopescu we checked two Mac builders with different versions and Radu's MacBook with the latest binary release from the download page; all of them cannot open `TFile::Open("https://...")`. I'll create a JIRA ticket.

**Comment by pcanal** [comment]
Can you rebase the branch (and fix potential conflicts)?

**Comment by jblomer** [comment]
@pcanal done.
